eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,342 @@
+		signer:   types.MakeSigner(api.backend.ChainConfig(), block.Number()),
+		state:    statedb,
+		tracer:   vm.NewStructLogger(config.LogConfig),
+		blockCtx: core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil),
+		StorageTrace: &types.StorageTrace{
+			RootBefore:    parent.Root(),
+			RootAfter:     block.Root(),
+			Proofs:        make(map[string][]hexutil.Bytes),
+			StorageProofs: make(map[string]map[string][]hexutil.Bytes),
+		},
+		executionResults: make([]*types.ExecutionResult, block.Transactions().Len()),
+	}
+
+	for _, coinbase := range []common.Address{block.Coinbase(), api.backend.Coinbase()} {
+		if coinbase == (common.Address{}) {
+			continue
+		}
+		key := coinbase.String()
+		if _, exist := env.Proofs[key]; !exist {
+			proof, err := env.state.GetProof(coinbase)
+			if err != nil {
+				log.Error("Proof for coinbase not available", "coinbase", coinbase, "error", err)
+				// but we still mark the proofs map with nil array
+			}
+			wrappedProof := make([]hexutil.Bytes, len(proof))
+			for i, bt := range proof {
+				wrappedProof[i] = bt
+			}
+			env.Proofs[key] = wrappedProof
+		}
+	}
+	return env, nil
+}
+
+func (api *API) getBlockResult(block *types.Block, env *environment) (*types.BlockResult, error) {
+	// Execute all the transaction contained within the block concurrently
+	var (
+		txs  = block.Transactions()
+		pend = new(sync.WaitGroup)
+		jobs = make(chan *txTraceTask, len(txs))
+	)
+	threads := runtime.NumCPU()
+	if threads > len(txs) {
+		threads = len(txs)
+	}
+	for th := 0; th < threads; th++ {
+		pend.Add(1)
+		go func() {
+			defer pend.Done()
+			// Fetch and execute the next transaction trace tasks
+			for task := range jobs {
+				if err := api.traceTx2(env, task.statedb, task.index, block); err != nil {
+					log.Error("failed to trace tx", "txHash", txs[task.index].Hash().String())
+				}
+			}
+		}()
+	}
+
+	// Feed the transactions into the tracers and return
+	var failed error
+	for i, tx := range txs {
+		// Send the trace task over for execution
+		jobs <- &txTraceTask{statedb: env.state.Copy(), index: i}
+
+		// Generate the next state snapshot fast without tracing
+		msg, _ := tx.AsMessage(env.signer, block.BaseFee())
+		env.state.Prepare(tx.Hash(), i)
+		vmenv := vm.NewEVM(env.blockCtx, core.NewEVMTxContext(msg), env.state, api.backend.ChainConfig(), vm.Config{})
+		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas())); err != nil {
+			failed = err
+			break
+		}
+		// Finalize the state so any modifications are written to the trie
+		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
+		env.state.Finalise(vmenv.ChainConfig().IsEIP158(block.Number()))
+	}
+	close(jobs)
+	pend.Wait()
+
+	// If execution failed in between, abort
+	if failed != nil {
+		return nil, failed
+	}
+
+	return api.writeBlockResult(env, block)
+}
+
+func (api *API) traceTx2(env *environment, state *state.StateDB, index int, block *types.Block) error {
```

Comment:
why "2" in the func?

Reply:
It's not that much normative, already changed.